pkg/async/schedule/flytescheduler/go_gf.go

```diff
@@ -0,0 +1,137 @@
+package flytescheduler
+
+import (
+	"context"
+	"fmt"
+	"github.com/flyteorg/flyteadmin/pkg/async/schedule/aws"
+	"github.com/flyteorg/flyteadmin/pkg/repositories/models"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/gogf/gf/os/gcron"
+	"github.com/gogf/gf/os/gtimer"
+	"github.com/robfig/cron"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type GoGF struct {
+}
+
+func (g GoGF) GetScheduleName(s models.SchedulableEntity) string {
+	return strconv.FormatUint(aws.HashIdentifier(core.Identifier{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    s.Name,
+		Version: s.Version,
+	}), 10)
+}
+
+func (g GoGF) Register(ctx context.Context, s models.SchedulableEntity, funcRef func()) error {
+	nameOfSchedule := g.GetScheduleName(s)
+
+	if s.Active {
+		if len(s.CronExpression) > 0 {
+			err := addCronJob(s.CronExpression,funcRef,nameOfSchedule)
+			if err != nil {
+				logger.Errorf(ctx, "failed to add cron schedule %v due to %v", s, err)
+			}
+		} else {
+			err := addFixedIntervalJob(s.Unit, s.FixedRateValue, funcRef)
+			if err != nil {
+				logger.Errorf(ctx, "failed to add fixed rate schedule %v due to %v", s, err)
+			}
+		}
+	} else {
+		if len(s.CronExpression) > 0 {
+			removeCronJob(ctx, nameOfSchedule)
+		} else {
+			removeFixedIntervalJob(ctx, nameOfSchedule)
+		}
+	}
+	return nil
+}
+
+func (g GoGF) GetScheduledTime(s models.SchedulableEntity, fromTime time.Time) (time.Time, error) {
+	if len(s.CronExpression) > 0 {
+		return getCronScheduledTime(s.CronExpression, fromTime)
+	} else {
+		return getFixedIntervalScheduledTime(s.Unit, s.FixedRateValue, fromTime)
+	}
+}
+
+func (g GoGF) GetCatchUpTimes(schedule models.SchedulableEntity, from time.Time, to time.Time) ([]time.Time, error){
+	var scheduledTimes []time.Time
+	currFrom := from
+	for currFrom.Before(to) {
+		scheduledTime, err := g.GetScheduledTime(schedule, currFrom)
+		if err != nil {
+			return nil, err
+		}
+		scheduledTimes = append(scheduledTimes, scheduledTime)
+		currFrom = scheduledTime
+	}
+	return scheduledTimes, nil
+}
+
+func addCronJob(cronExpression string, job func(), nameOfSchedule string) error {
+	_, err := gcron.AddSingleton(cronExpression, job, nameOfSchedule)
+	if err != nil && strings.HasSuffix(err.Error(), "already exists"){
```

Comment:
this seems brittle, is there no code or anything exposed by the library?

Reply:
Sure. will use gerror.CodeInvalidOperation https://github.com/gogf/gf/blob/e9ea58df64c297f89eec633e3ee4eda4b290c8f1/os/gcron/gcron_cron.go#L65

Reply:
This change just got  added in latest version 15 days ago so using the latest to use error code now

Reply:
Done